### PR TITLE
esp8266: Improved time keeping

### PR DESCRIPTION
**Pull request by Adam5Wu**
1. use more stable system clock as general time source;
2. increase rtc accuracy and reduce drifts.

### CODE REVIEW

**Comment by Adam5Wu** [comment]
Problem reference: #2724

File: esp8266/machine_rtc.c [hunk 1]
```diff
@@ -55,23 +56,96 @@ uint64_t pyb_rtc_alarm0_expiry; // in microseconds
...
+    uint32_t clk_ticks;
+
+    offset = clk_offset;
+    clk_ticks = system_get_time();
+
+    if (clk_ticks >= clk_last_ticks) {
+      offset+= clk_ticks-clk_last_ticks;
+    } else {
+      // If overflow happened, assume 1 wrap-around and persist info for the new cycle
+      offset+= clk_ticks+~clk_last_ticks+1;
```
Comment (dpgeorge) [inline esp8266/machine_rtc.c]: Did you see that there is system_time_high_word already defined elsewhere which handles overflow?  See mp_hal_ticks_ms() for example.  It would be good to reuse this counter, then probably you don't need esp_clk_XXX functions.

File: esp8266/machine_rtc.c [hunk 2]
```diff
@@ -55,23 +56,96 @@ uint64_t pyb_rtc_alarm0_expiry; // in microseconds
...
+      clk_last_ticks = clk_ticks;
+      clk_offset = offset;
+    }
+    return offset; 
+};
+
+void pyb_rtc_set_us_since_2000(uint64_t nowus) {
```
Comment (dpgeorge) [inline esp8266/machine_rtc.c]: Please don't (unnecessarily) move the definition of these functions, it makes it hard to see exactly how they changed.

File: esp8266/machine_rtc.c [hunk 3]
```diff
@@ -151,14 +193,15 @@ STATIC mp_obj_t pyb_rtc_datetime(mp_uint_t n_args, const mp_obj_t *args) {
...
-        pyb_rtc_set_us_since_2000(
-            ((uint64_t)timeutils_seconds_since_2000(
-                mp_obj_get_int(items[0]),
-                mp_obj_get_int(items[1]),
-                mp_obj_get_int(items[2]),
-                mp_obj_get_int(items[4]),
-                mp_obj_get_int(items[5]),
-                mp_obj_get_int(items[6])) * 1000 + mp_obj_get_int(items[7])) * 1000);
+        uint64_t arg_us = ((uint64_t)timeutils_seconds_since_2000(
+                               mp_obj_get_int(items[0]),
```
Comment (dpgeorge) [inline esp8266/machine_rtc.c]: Please keep original indentation of the lines that don't change.

File: esp8266/machine_rtc.c (hunk 1 above)
Comment (Adam5Wu) [inline esp8266/machine_rtc.c]: I saw that one. However, I think that variable is for slightly different purposes.
The "system clock" is designed to be adjustable, i.e. if for any reason a program wants to adjust current system time, the "offset" will be amended accordingly;
The system_time_high_word is for tracking the time differences between two ref points, and adjusting system clock should not affect this counter, otherwise the result will be wrong after adjusting the system clock.

File: esp8266/machine_rtc.c (hunk 2 above)
Comment (Adam5Wu) [inline esp8266/machine_rtc.c]: I have move the pyb_rtc... functions here because I would like to use the set function in mp_hal_rtc_init(), and I think having the definition before use make make code reading a little easier. :)
The new pyb_rtc_... function are completely rewritten, so they have little relation to the old code. Their code logic are more like the esp_clk_... functions.

File: esp8266/machine_rtc.c (hunk 1 above)
Comment (Adam5Wu) [inline esp8266/machine_rtc.c]: I have made a new revision, which reuses system_time_high_word for time keeping, and just maintain an offset of "time between 2000 and last reboot" for system clock.

It does not eliminate the esp_clk_... functions, but made it a lot simpler. :D

**Comment by jcea** [comment]
Any news about integration of this PR?.

**Comment by pfalcon** [comment]
This PR changes more things than needed, so it's hard even to review, so nobody did that so far (in full). Beyond that, it needs a lot of testing, because otherwise it may turn out that it fixes one thing and break others. So @jcea, if you've done an extended testing of this patch, please report details of it (what exactly was tested, how, for how long, etc.).

**Comment by pfalcon** [comment]
Oh, and there're still code style issues. (Which always rings a bell - if there're errors is such a simple and visible matter as punctuation, what more complex, less visible errors the code may contain?)

**Comment by Adam5Wu** [comment]
@pfalcon, I think by saying "punctuation" style issue, you really meant to say "indentation", as already fixed in eb0ad96? Or could you please be more specific on the "punctuation" you are referring to?

Also, would you care to point out what unecessary code is involved in your opinion?

Without specifics, it is hard to follow your hand-waving arguments to the conclusion you made.

Or maybe your conclusion is simply infallible:
Yes, to err is to human, so who knows what can go wrong, I didn't check so anything is possible

**Comment by Adam5Wu** [comment]
Oh well, this is getting absurd.

For the record, I have switched to Arduino esp8266 months ago, because I found the current uP implementation for esp8266 cannot meet my expectation of usability and stability. I decided to give it more time to mature before I venture back again.

I kept my fork and PRs with the good will of at least save other people's effort and time I have spent on solving these problems. But clearly, it has become more of an eyesore to some maintainers.

So, one less problem for you, one less problem for me. :P

**Comment by PvdBerg1998** [comment]
Closing this PR is very unfortunate as I'm hitting the same issue for a time critical application. @pfalcon is this PR still applicable?
